packages/modules/event-bus-redis/src/services/event-bus-redis.ts

```diff
@@ -71,78 +81,114 @@ export default class RedisEventBusService extends AbstractEventBusModuleService
+  async emit<T = unknown>(
+    eventsData: Message<T> | Message<T>[],
+    options: BulkJobOptions = {}
+  ): Promise<void> {
+    let eventsDataArray = Array.isArray(eventsData) ? eventsData : [eventsData]
 
-  async emit<T>(data: Message<T>[]): Promise<void>
+    const eventsToEmit = eventsDataArray.filter(
+      (eventData) => !isPresent(eventData.body?.metadata?.eventGroupId)
+    )
 
-  async emit<T, TInput extends string | EmitData<T>[] | Message<T>[] = string>(
-    eventNameOrData: TInput,
-    data?: T,
-    options: BulkJobOptions | JobsOptions = {}
-  ): Promise<void> {
-    const globalJobOptions = this.moduleOptions_.jobOptions ?? {}
+    const eventsToGroup = eventsDataArray.filter((eventData) =>
+      isPresent(eventData.body?.metadata?.eventGroupId)
+    )
 
-    const isBulkEmit = Array.isArray(eventNameOrData)
+    const groupEventsMap = new Map<string, Message<T>[]>()
 
-    const opts = {
-      // default options
-      removeOnComplete: true,
-      attempts: 1,
-      // global options
-      ...globalJobOptions,
-    } as EmitOptions
-
-    const dataBody = isString(eventNameOrData)
-      ? data ?? (data as Message<T>).body
-      : undefined
-
-    const events = isBulkEmit
-      ? eventNameOrData.map((event) => ({
-          name: event.eventName,
-          data: {
-            eventName: event.eventName,
-            data: (event as EmitData).data ?? (event as Message<T>).body,
-          },
-          opts: {
-            ...opts,
-            // local options
-            ...event.options,
-          },
-        }))
-      : [
-          {
-            name: eventNameOrData as string,
-            data: { eventName: eventNameOrData, data: dataBody },
-            opts: {
-              ...opts,
-              // local options
-              ...options,
-            },
-          },
-        ]
-
-    await this.queue_.addBulk(events)
+    for (const event of eventsToGroup) {
+      const groupId = event.body.metadata?.eventGroupId!
+      const array = groupEventsMap.get(groupId) ?? []
+
+      array.push(event)
+      groupEventsMap.set(groupId, array)
+    }
+
+    const promises: Promise<unknown>[] = []
+
+    if (eventsToEmit.length) {
+      const emitData = this.buildEvents(eventsToEmit, options)
+
+      promises.push(this.queue_.addBulk(emitData))
+    }
+
+    for (const [groupId, events] of groupEventsMap.entries()) {
+      const eventsData = this.buildEvents(events, options)
+
+      promises.push(this.groupEvents(groupId, eventsData))
+    }
+
+    await promiseAll(promises)
   }
 
-  // TODO: Implement redis based staging + release
-  async releaseGroupedEvents(eventGroupId: string) {}
-  async clearGroupedEvents(eventGroupId: string) {}
+  private async groupEvents<T = unknown>(
+    eventGroupId: string,
+    events: IORedisEventType<T>[]
+  ) {
+    await this.eventBusRedisConnection_.rpush(
+      `staging:${eventGroupId}`,
+      ...events.map((event) => JSON.stringify(event))
+    )
```

Comment:
thought: Wondering if we should have a TTL on this

Reply:
At the very least maybe an overview in the admin (similar to workflows) that you can action.

Reply:
@riqwan I was thinking it would be nice to be able to CRUD event groups from the UI, so you can see what's staged, what is maybe stuck, etc. It is orthogonal to TTLs, I was just throwing it out as an idea :)

Reply:
I am not sure the ui should be handled by us right? They already have tools for redis and bull to monitor and see this directly

Reply:
I was just thinking having this embedded experience similar to the workflows UI would be nice, esp. when it comes to grouped events, but of course it's not something that is pressing to implement

Reply:
Somewhat related, I have always found Stripe's Events API (and the related UI) very nice and super useful for debugging. A feature like this would probably need its own module, and the events persisted to the DB, so a bit different from this. But maybe a combination could make sense in the future 😄